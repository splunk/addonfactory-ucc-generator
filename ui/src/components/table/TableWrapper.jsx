import React, { useState, useContext, useEffect, memo } from 'react';
import update from 'immutability-helper';
import axios from 'axios';
import PropTypes from 'prop-types';

import { WaitSpinnerWrapper } from './CustomTableStyle';
import { axiosCallWrapper } from '../../util/axiosCallWrapper';
import { getUnifiedConfigs, generateToast, isTrue } from '../../util/util';
import CustomTable from './CustomTable';
import TableHeader from './TableHeader';
import TableContext from '../../context/TableContext';
import { PAGE_INPUT } from '../../constants/pages';

function TableWrapper({
    page,
    serviceName,
    handleRequestModalOpen,
    handleOpenPageStyleDialog,
    displayActionBtnAllRows,
}) {
    const [sortKey, setSortKey] = useState('name');
    const [sortDir, setSortDir] = useState('asc');
    const [loading, setLoading] = useState(true);
    const [error, setError] = useState(null);

    const { rowData, setRowData, pageSize, currentPage, searchText, searchType } =
        useContext(TableContext);

    const unifiedConfigs = getUnifiedConfigs();

    const services =
        page === PAGE_INPUT
            ? unifiedConfigs.pages.inputs.services
            : unifiedConfigs.pages.configuration.tabs.filter((x) => x.name === serviceName);

    const tableConfig =
        page === PAGE_INPUT
            ? unifiedConfigs.pages.inputs.table ||
              services.find((x) => x.name === serviceName).table
            : unifiedConfigs.pages.configuration.tabs.find((x) => x.name === serviceName).table;

    const { moreInfo } = tableConfig;
    const headers = tableConfig.header;
    const isTabs = !!serviceName;

    const modifyAPIResponse = (data) => {
        const obj = {};
        services.forEach((service, index) => {
            if (service && service.name && data) {
                const tmpObj = {};
                data[index].forEach((val) => {
                    tmpObj[val.name] = {
                        ...val.content,
                        id: val.id,
                        name: val.name,
                        serviceName: service.name,
                        serviceTitle: service.title || '',
                    };
                });
                obj[service.name] = tmpObj;
            }
        });
        setRowData(obj);
        setLoading(false);
    };

    const fetchInputs = () => {
        const requests = [];
        services.forEach((service) => {
            requests.push(
                axiosCallWrapper({
                    serviceName: service.name,
                    params: { count: -1 },
                })
            );
        });
        axios
            .all(requests)
            // eslint-disable-next-line no-shadow
            .catch((error) => {
                let message = '';
                let errorCode = '';
                if (error.response) {
                    // The request was made and the server responded with a status code
                    message = `Error received from server: ${error.response.data.messages[0].text}`;
                    errorCode = page === PAGE_INPUT ? 'ERR0001' : 'ERR0002';
                } else if (error.request) {
                    // The request was made but no response was received
                    message = `No response received while making request to ${page} services`;
                    errorCode = 'ERR0003';
                } else {
                    // Something happened in setting up the request that triggered an Error
                    message = `Error making request to ${page} services`;
                    errorCode = 'ERR0004';
                }
                // eslint-disable-next-line no-param-reassign
                error.uccErrorCode = errorCode;
                generateToast(message);
                setLoading(false);
                setError(error);
                return Promise.reject(error);
            })
            .then((response) => {
                modifyAPIResponse(response.map((res) => res.data.entry));
            });
    };

    useEffect(() => {
        fetchInputs();
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, []);

    /**
     *
     * @param row {Object} row
     */
    const changeToggleStatus = (row) => {
        setRowData((currentRowData) =>
            update(currentRowData, {
                [row.serviceName]: {
                    [row.name]: {
                        __toggleShowSpinner: { $set: true },
                    },
                },
            })
        );
        const body = new URLSearchParams();
        body.append('disabled', !row.disabled);
        axiosCallWrapper({
            serviceName: `${row.serviceName}/${row.name}`,
            body,
            customHeaders: { 'Content-Type': 'application/x-www-form-urlencoded' },
            method: 'post',
            handleError: true,
            callbackOnError: () => {
                setRowData((currentRowData) =>
                    update(currentRowData, {
                        [row.serviceName]: {
                            [row.name]: {
                                __toggleShowSpinner: { $set: false },
                            },
                        },
                    })
                );
            },
        }).then((response) => {
            setRowData((currentRowData) =>
                update(currentRowData, {
                    [row.serviceName]: {
                        [row.name]: {
                            // ADDON-39125: isTrue required if splunktaucclib resthandlers' super() is not invoked
                            disabled: { $set: isTrue(response.data.entry[0].content.disabled) },
                            __toggleShowSpinner: { $set: false },
                        },
                    },
                })
            );
        });
    };

    const handleSort = (e, val) => {
        const prevSortKey = sortKey;
        const prevSortDir = prevSortKey === val.sortKey ? sortDir : 'none';
        const nextSortDir = prevSortDir === 'asc' ? 'desc' : 'asc';
        setSortDir(nextSortDir);
        setSortKey(val.sortKey);
    };

    /**
     *
     * @param {Array} data
     * This function will iterate an arrray and match each key-value with the searchText
     * It will return a new array which will match with searchText
     */
    const findByMatchingValue = (data) => {
        const arr = [];
        const tableFields = [];

        headers.forEach((headData) => {
            tableFields.push(headData.field);
        });
        moreInfo?.forEach((moreInfoData) => {
            tableFields.push(moreInfoData.field);
        });

        Object.keys(data).forEach((v) => {
            let found = false;
            Object.keys(data[v]).forEach((vv) => {
                if (
                    tableFields.includes(vv) &&
                    typeof data[v][vv] === 'string' &&
                    data[v][vv].toLowerCase().includes(searchText.toLowerCase().trim()) &&
                    !found
                ) {
                    arr.push(data[v]);
                    found = true;
                }
            });
        });
        return arr;
    };

    const getRowData = () => {
        let arr = [];
        if (searchType === 'all') {
            Object.keys(rowData).forEach((key) => {
                let newArr = [];
                if (searchText && searchText.length) {
                    newArr = findByMatchingValue(rowData[key]);
                } else {
                    newArr = Object.keys(rowData[key]).map((val) => rowData[key][val]);
                }
                arr = arr.concat(newArr);
            });
        } else {
            arr = findByMatchingValue(rowData[searchType]);
        }

        // For Inputs page, filter the data when tab change
        if (isTabs) {
            arr = arr.filter((v) => v.serviceName === serviceName);
        }

        // Sort the array based on the sort value
        const sortedArr = arr.sort((rowA, rowB) => {
            if (sortDir === 'asc') {
                const rowAValue = rowA[sortKey] === undefined ? '' : rowA[sortKey];
                const rowBValue = rowB[sortKey] === undefined ? '' : rowB[sortKey];
                return rowAValue > rowBValue ? 1 : -1;
            }
            if (sortDir === 'desc') {
                const rowAValue = rowA[sortKey] === undefined ? '' : rowA[sortKey];
                const rowBValue = rowB[sortKey] === undefined ? '' : rowB[sortKey];
                return rowBValue > rowAValue ? 1 : -1;
            }
            return 0;
        });

        let updatedArr = sortedArr.slice(currentPage * pageSize, (currentPage + 1) * pageSize);

        if (currentPage > 0 && !updatedArr.length) {
            updatedArr = sortedArr.slice((currentPage - 1) * pageSize, pageSize);
        }

        return [updatedArr, arr.length, arr];
    };

    if (error?.uccErrorCode) {
        throw error;
    }

    if (loading) {
        return <WaitSpinnerWrapper size="medium" />;
    }

    const [filteredData, totalElement, allFilteredData] = getRowData();
    return (
        <>
            <TableHeader
                page={page}
                services={services}
                totalElement={totalElement}
                handleRequestModalOpen={handleRequestModalOpen}
                changeToggleStatus={changeToggleStatus}
                isTabs={isTabs}
                allFilteredData={allFilteredData}
                displayActionBtnAllRows={displayActionBtnAllRows}
            />
            <CustomTable
                page={page}
                serviceName={serviceName}
                data={filteredData}
                handleToggleActionClick={(row) => changeToggleStatus(row)}
                handleSort={handleSort}
                sortDir={sortDir}
                sortKey={sortKey}
                handleOpenPageStyleDialog={handleOpenPageStyleDialog}
                tableConfig={tableConfig}
                services={services}
            />
        </>
    );
}

TableWrapper.propTypes = {
    page: PropTypes.string,
    serviceName: PropTypes.string,
    handleRequestModalOpen: PropTypes.func,
    handleOpenPageStyleDialog: PropTypes.func,
    displayActionBtnAllRows: PropTypes.bool,
};

export default memo(TableWrapper);
