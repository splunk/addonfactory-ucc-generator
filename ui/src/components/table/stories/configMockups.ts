import { GlobalConfig } from '../../../types/globalConfig/globalConfig';

export const SIMPLE_NAME_TABLE_MOCK_DATA = {
    pages: {
        configuration: {
            tabs: [
                {
                    name: 'account',
                    table: {
                        actions: ['edit', 'delete', 'clone'],
                        header: [
                            {
                                label: 'Name',
                                field: 'name',
                            },
                        ],
                    },
                    entity: [
                        {
                            type: 'text',
                            label: 'Name',
                            validators: [
                                {
                                    type: 'string',
                                    errorMsg: 'Length of ID should be between 1 and 50',
                                    minLength: 1,
                                    maxLength: 50,
                                },
                                {
                                    type: 'regex',
                                    errorMsg:
                                        'Name must begin with a letter and consist exclusively of alphanumeric characters and underscores.',
                                    pattern: '^[a-zA-Z]\\w*$',
                                },
                            ],
                            field: 'name',
                            help: 'Enter a unique name for this account.',
                            required: true,
                        },
                    ],
                    title: 'Account',
                    restHandlerModule: 'splunk_ta_uccexample_validate_account_rh',
                    restHandlerClass: 'CustomAccountValidator',
                },
            ],
            title: 'Configuration',
            description: 'Set up your add-on',
        },
        inputs: {
            services: [
                {
                    name: 'example_input_one',
                    entity: [
                        {
                            type: 'text',
                            label: 'Name',
                            validators: [
                                {
                                    type: 'regex',
                                    errorMsg:
                                        'Input Name must begin with a letter and consist exclusively of alphanumeric characters and underscores.',
                                    pattern: '^[a-zA-Z]\\w*$',
                                },
                                {
                                    type: 'string',
                                    errorMsg: 'Length of input name should be between 1 and 100',
                                    minLength: 1,
                                    maxLength: 100,
                                },
                            ],
                            field: 'name',
                            help: 'A unique name for the data input.',
                            required: true,
                        },
                        {
                            type: 'checkbox',
                            label: 'Example Checkbox',
                            field: 'input_one_checkbox',
                            help: 'This is an example checkbox for the input one entity',
                            defaultValue: true,
                        },
                    ],
                    title: 'Example Input One',
                },
            ],
            title: 'Inputs',
            description: 'Manage your data inputs',
            table: {
                actions: ['edit', 'enable', 'delete', 'search', 'clone'],
                header: [
                    {
                        label: 'Name',
                        field: 'name',
                    },
                ],
                moreInfo: [
                    {
                        label: 'Name',
                        field: 'name',
                    },
                ],
            },
        },
    },
    meta: {
        name: 'Splunk_TA_UCCExample',
        restRoot: 'splunk_ta_uccexample',
        version: '5.41.0R9c5fbfe0',
        displayName: 'Splunk UCC test Add-on',
        schemaVersion: '0.0.3',
    },
} satisfies GlobalConfig;

export const getSimpleConfig = () => {
    const configCp = JSON.parse(JSON.stringify(SIMPLE_NAME_TABLE_MOCK_DATA));
    return configCp;
};

export const TABLE_CONFIG_WITH_MAPPING = {
    pages: {
        configuration: {
            tabs: [
                {
                    name: 'account',
                    table: {
                        actions: ['edit', 'delete', 'clone'],
                        header: [
                            {
                                label: 'Name',
                                field: 'name',
                            },
                            {
                                field: 'custom_text',
                                label: 'Custom Text',
                                mapping: {
                                    a: 'wxyz=a',
                                    ab: 'xyz=ab',
                                    abc: 'yz=abc',
                                    abcd: 'z=abcd',
                                },
                            },
                        ],
                    },
                    entity: [
                        {
                            type: 'text',
                            label: 'Name',
                            validators: [
                                {
                                    type: 'string',
                                    errorMsg: 'Length of ID should be between 1 and 50',
                                    minLength: 1,
                                    maxLength: 50,
                                },
                                {
                                    type: 'regex',
                                    errorMsg:
                                        'Name must begin with a letter and consist exclusively of alphanumeric characters and underscores.',
                                    pattern: '^[a-zA-Z]\\w*$',
                                },
                            ],
                            field: 'name',
                            help: 'Enter a unique name for this account.',
                            required: true,
                        },
                        {
                            type: 'text',
                            label: 'Custom Text',
                            field: 'custom_text',
                            help: 'custom text.',
                        },
                    ],
                    title: 'Account',
                    restHandlerModule: 'splunk_ta_uccexample_validate_account_rh',
                    restHandlerClass: 'CustomAccountValidator',
                },
            ],
            title: 'Configuration',
            description: 'Set up your add-on',
        },
        inputs: {
            services: [
                {
                    name: 'example_input_one',
                    entity: [
                        {
                            type: 'text',
                            label: 'Name',
                            validators: [
                                {
                                    type: 'regex',
                                    errorMsg:
                                        'Input Name must begin with a letter and consist exclusively of alphanumeric characters and underscores.',
                                    pattern: '^[a-zA-Z]\\w*$',
                                },
                                {
                                    type: 'string',
                                    errorMsg: 'Length of input name should be between 1 and 100',
                                    minLength: 1,
                                    maxLength: 100,
                                },
                            ],
                            field: 'name',
                            help: 'A unique name for the data input.',
                            required: true,
                        },
                        {
                            type: 'checkbox',
                            label: 'Example Checkbox',
                            field: 'input_one_checkbox',
                            help: 'This is an example checkbox for the input one entity',
                            defaultValue: true,
                        },
                    ],
                    title: 'Example Input One',
                },
            ],
            title: 'Inputs',
            description: 'Manage your data inputs',
            table: {
                actions: ['edit', 'enable', 'delete', 'search', 'clone'],
                header: [
                    {
                        label: 'Name',
                        field: 'name',
                    },
                ],
                moreInfo: [
                    {
                        label: 'Name',
                        field: 'name',
                    },
                ],
            },
        },
    },
    meta: {
        name: 'Splunk_TA_UCCExample',
        restRoot: 'splunk_ta_uccexample',
        version: '5.41.0R9c5fbfe0',
        displayName: 'Splunk UCC test Add-on',
        schemaVersion: '0.0.3',
    },
} satisfies GlobalConfig;
