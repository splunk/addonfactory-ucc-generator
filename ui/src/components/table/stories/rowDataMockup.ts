import { HttpResponse, http } from 'msw';

export const ROW_DATA = [
    {
        name: 'aaaaaa',
        id: 'https://localhost:8000/servicesNS/nobody/Splunk_TA_UCCExample/splunk_ta_uccexample_account/aaaaaa',
        updated: '1970-01-01T00:00:00+00:00',
        links: {
            alternate:
                '/servicesNS/nobody/Splunk_TA_UCCExample/splunk_ta_uccexample_account/aaaaaa',
            list: '/servicesNS/nobody/Splunk_TA_UCCExample/splunk_ta_uccexample_account/aaaaaa',
            edit: '/servicesNS/nobody/Splunk_TA_UCCExample/splunk_ta_uccexample_account/aaaaaa',
            remove: '/servicesNS/nobody/Splunk_TA_UCCExample/splunk_ta_uccexample_account/aaaaaa',
        },
        author: 'admin',
        acl: {
            app: 'Splunk_TA_UCCExample',
            can_change_perms: true,
            can_list: true,
            can_share_app: true,
            can_share_global: true,
            can_share_user: true,
            can_write: true,
            modifiable: true,
            owner: 'admin',
            perms: {
                read: ['*'],
                write: ['admin', 'sc_admin'],
            },
            removable: true,
            sharing: 'global',
        },
        content: {
            account_multiple_select: 'one',
            account_radio: '1',
            auth_type: 'basic',
            custom_endpoint: 'login.example.com',
            disabled: false,
            'eai:acl': null,
            'eai:appName': 'Splunk_TA_UCCExample',
            'eai:userName': 'nobody',
            password: '******',
            custom_text: 'a',
            token: '******',
            username: 'aaaaaa',
        },
    },
    {
        name: 'aaaaaaa',
        id: 'https://localhost:8000/servicesNS/nobody/Splunk_TA_UCCExample/splunk_ta_uccexample_account/aaaaaaa',
        updated: '1970-01-01T00:00:00+00:00',
        links: {
            alternate:
                '/servicesNS/nobody/Splunk_TA_UCCExample/splunk_ta_uccexample_account/aaaaaaa',
            list: '/servicesNS/nobody/Splunk_TA_UCCExample/splunk_ta_uccexample_account/aaaaaaa',
            edit: '/servicesNS/nobody/Splunk_TA_UCCExample/splunk_ta_uccexample_account/aaaaaaa',
            remove: '/servicesNS/nobody/Splunk_TA_UCCExample/splunk_ta_uccexample_account/aaaaaaa',
        },
        author: 'admin',
        acl: {
            app: 'Splunk_TA_UCCExample',
            can_change_perms: true,
            can_list: true,
            can_share_app: true,
            can_share_global: true,
            can_share_user: true,
            can_write: true,
            modifiable: true,
            owner: 'admin',
            perms: {
                read: ['*'],
                write: ['admin', 'sc_admin'],
            },
            removable: true,
            sharing: 'global',
        },
        content: {
            account_multiple_select: 'one',
            account_radio: '1',
            auth_type: 'basic',
            custom_endpoint: 'login.example.com',
            disabled: false,
            'eai:acl': null,
            'eai:appName': 'Splunk_TA_UCCExample',
            'eai:userName': 'nobody',
            password: '******',
            custom_text: 'ab',
            token: '******',
            username: 'aaaaaaa',
        },
    },
    {
        name: 'bbbb',
        id: 'https://localhost:8000/servicesNS/nobody/Splunk_TA_UCCExample/splunk_ta_uccexample_account/bbbb',
        updated: '1970-01-01T00:00:00+00:00',
        links: {
            alternate: '/servicesNS/nobody/Splunk_TA_UCCExample/splunk_ta_uccexample_account/bbbb',
            list: '/servicesNS/nobody/Splunk_TA_UCCExample/splunk_ta_uccexample_account/bbbb',
            edit: '/servicesNS/nobody/Splunk_TA_UCCExample/splunk_ta_uccexample_account/bbbb',
            remove: '/servicesNS/nobody/Splunk_TA_UCCExample/splunk_ta_uccexample_account/bbbb',
        },
        author: 'admin',
        acl: {
            app: 'Splunk_TA_UCCExample',
            can_change_perms: true,
            can_list: true,
            can_share_app: true,
            can_share_global: true,
            can_share_user: true,
            can_write: true,
            modifiable: true,
            owner: 'admin',
            perms: {
                read: ['*'],
                write: ['admin', 'sc_admin'],
            },
            removable: true,
            sharing: 'global',
        },
        content: {
            account_multiple_select: 'one',
            account_radio: '1',
            auth_type: 'basic',
            custom_endpoint: 'login.example.com',
            disabled: false,
            'eai:acl': null,
            'eai:appName': 'Splunk_TA_UCCExample',
            'eai:userName': 'nobody',
            password: '******',
            custom_text: 'abc',
            token: '******',
            username: 'aaaaaa',
        },
    },
    {
        name: 'ccc',
        id: 'https://localhost:8000/servicesNS/nobody/Splunk_TA_UCCExample/splunk_ta_uccexample_account/ccc',
        updated: '1970-01-01T00:00:00+00:00',
        links: {
            alternate: '/servicesNS/nobody/Splunk_TA_UCCExample/splunk_ta_uccexample_account/ccc',
            list: '/servicesNS/nobody/Splunk_TA_UCCExample/splunk_ta_uccexample_account/ccc',
            edit: '/servicesNS/nobody/Splunk_TA_UCCExample/splunk_ta_uccexample_account/ccc',
            remove: '/servicesNS/nobody/Splunk_TA_UCCExample/splunk_ta_uccexample_account/ccc',
        },
        author: 'admin',
        acl: {
            app: 'Splunk_TA_UCCExample',
            can_change_perms: true,
            can_list: true,
            can_share_app: true,
            can_share_global: true,
            can_share_user: true,
            can_write: true,
            modifiable: true,
            owner: 'admin',
            perms: {
                read: ['*'],
                write: ['admin', 'sc_admin'],
            },
            removable: true,
            sharing: 'global',
        },
        content: {
            account_multiple_select: 'one',
            account_radio: '1',
            auth_type: 'basic',
            custom_endpoint: 'login.example.com',
            disabled: false,
            'eai:acl': null,
            'eai:appName': 'Splunk_TA_UCCExample',
            'eai:userName': 'nobody',
            password: '******',
            custom_text: 'abcd',
            token: '******',
            username: 'ccc',
        },
    },
    {
        name: 'ddddd',
        id: 'https://localhost:8000/servicesNS/nobody/Splunk_TA_UCCExample/splunk_ta_uccexample_account/ddddd',
        updated: '1970-01-01T00:00:00+00:00',
        links: {
            alternate: '/servicesNS/nobody/Splunk_TA_UCCExample/splunk_ta_uccexample_account/ddddd',
            list: '/servicesNS/nobody/Splunk_TA_UCCExample/splunk_ta_uccexample_account/ddddd',
            edit: '/servicesNS/nobody/Splunk_TA_UCCExample/splunk_ta_uccexample_account/ddddd',
            remove: '/servicesNS/nobody/Splunk_TA_UCCExample/splunk_ta_uccexample_account/ddddd',
        },
        author: 'admin',
        acl: {
            app: 'Splunk_TA_UCCExample',
            can_change_perms: true,
            can_list: true,
            can_share_app: true,
            can_share_global: true,
            can_share_user: true,
            can_write: true,
            modifiable: true,
            owner: 'admin',
            perms: {
                read: ['*'],
                write: ['admin', 'sc_admin'],
            },
            removable: true,
            sharing: 'global',
        },
        content: {
            account_multiple_select: 'two',
            account_radio: '1',
            auth_type: 'basic',
            custom_endpoint: 'login.example.com',
            disabled: false,
            'eai:acl': null,
            'eai:appName': 'Splunk_TA_UCCExample',
            'eai:userName': 'nobody',
            password: '******',
            custom_text: 'ab',
            token: '******',
            username: 'ddddd',
        },
    },
    {
        name: 'test1',
        id: 'https://localhost:8000/servicesNS/nobody/Splunk_TA_UCCExample/splunk_ta_uccexample_account/test1',
        updated: '1970-01-01T00:00:00+00:00',
        links: {
            alternate: '/servicesNS/nobody/Splunk_TA_UCCExample/splunk_ta_uccexample_account/test1',
            list: '/servicesNS/nobody/Splunk_TA_UCCExample/splunk_ta_uccexample_account/test1',
            edit: '/servicesNS/nobody/Splunk_TA_UCCExample/splunk_ta_uccexample_account/test1',
            remove: '/servicesNS/nobody/Splunk_TA_UCCExample/splunk_ta_uccexample_account/test1',
        },
        author: 'admin',
        acl: {
            app: 'Splunk_TA_UCCExample',
            can_change_perms: true,
            can_list: true,
            can_share_app: true,
            can_share_global: true,
            can_share_user: true,
            can_write: true,
            modifiable: true,
            owner: 'admin',
            perms: {
                read: ['*'],
                write: ['admin', 'sc_admin'],
            },
            removable: true,
            sharing: 'global',
        },
        content: {
            account_multiple_select: 'two',
            account_radio: '1',
            auth_type: 'basic',
            custom_endpoint: 'login.example.com',
            disabled: false,
            'eai:acl': null,
            'eai:appName': 'Splunk_TA_UCCExample',
            'eai:userName': 'nobody',
            password: '******',
            custom_text: 'aaaaa',
            token: '******',
            username: 'test1',
        },
    },
    {
        name: 'test2',
        id: 'https://localhost:8000/servicesNS/nobody/Splunk_TA_UCCExample/splunk_ta_uccexample_account/test2',
        updated: '1970-01-01T00:00:00+00:00',
        links: {
            alternate: '/servicesNS/nobody/Splunk_TA_UCCExample/splunk_ta_uccexample_account/test2',
            list: '/servicesNS/nobody/Splunk_TA_UCCExample/splunk_ta_uccexample_account/test2',
            edit: '/servicesNS/nobody/Splunk_TA_UCCExample/splunk_ta_uccexample_account/test2',
            remove: '/servicesNS/nobody/Splunk_TA_UCCExample/splunk_ta_uccexample_account/test2',
        },
        author: 'admin',
        acl: {
            app: 'Splunk_TA_UCCExample',
            can_change_perms: true,
            can_list: true,
            can_share_app: true,
            can_share_global: true,
            can_share_user: true,
            can_write: true,
            modifiable: true,
            owner: 'admin',
            perms: {
                read: ['*'],
                write: ['admin', 'sc_admin'],
            },
            removable: true,
            sharing: 'global',
        },
        content: {
            account_multiple_select: 'two',
            account_radio: '1',
            auth_type: 'basic',
            custom_endpoint: 'login.example.com',
            disabled: false,
            'eai:acl': null,
            'eai:appName': 'Splunk_TA_UCCExample',
            'eai:userName': 'nobody',
            password: '******',
            custom_text: 'two',
            token: '******',
            username: 'test1',
        },
    },
    {
        name: 'testsomethingelse',
        id: 'https://localhost:8000/servicesNS/nobody/Splunk_TA_UCCExample/splunk_ta_uccexample_account/testsomethingelse',
        updated: '1970-01-01T00:00:00+00:00',
        links: {
            alternate:
                '/servicesNS/nobody/Splunk_TA_UCCExample/splunk_ta_uccexample_account/testsomethingelse',
            list: '/servicesNS/nobody/Splunk_TA_UCCExample/splunk_ta_uccexample_account/testsomethingelse',
            edit: '/servicesNS/nobody/Splunk_TA_UCCExample/splunk_ta_uccexample_account/testsomethingelse',
            remove: '/servicesNS/nobody/Splunk_TA_UCCExample/splunk_ta_uccexample_account/testsomethingelse',
        },
        author: 'admin',
        acl: {
            app: 'Splunk_TA_UCCExample',
            can_change_perms: true,
            can_list: true,
            can_share_app: true,
            can_share_global: true,
            can_share_user: true,
            can_write: true,
            modifiable: true,
            owner: 'admin',
            perms: {
                read: ['*'],
                write: ['admin', 'sc_admin'],
            },
            removable: true,
            sharing: 'global',
        },
        content: {
            account_multiple_select: 'two',
            account_radio: '1',
            auth_type: 'basic',
            custom_endpoint: 'login.example.com',
            disabled: false,
            'eai:acl': null,
            'eai:appName': 'Splunk_TA_UCCExample',
            'eai:userName': 'nobody',
            password: '******',
            custom_text: 'testsomethingelse',
            token: '******',
            username: 'test1',
        },
    },
    {
        name: 'zzzzzzz',
        id: 'https://localhost:8000/servicesNS/nobody/Splunk_TA_UCCExample/splunk_ta_uccexample_account/zzzzzzz',
        updated: '1970-01-01T00:00:00+00:00',
        links: {
            alternate:
                '/servicesNS/nobody/Splunk_TA_UCCExample/splunk_ta_uccexample_account/zzzzzzz',
            list: '/servicesNS/nobody/Splunk_TA_UCCExample/splunk_ta_uccexample_account/zzzzzzz',
            edit: '/servicesNS/nobody/Splunk_TA_UCCExample/splunk_ta_uccexample_account/zzzzzzz',
            remove: '/servicesNS/nobody/Splunk_TA_UCCExample/splunk_ta_uccexample_account/zzzzzzz',
        },
        author: 'admin',
        acl: {
            app: 'Splunk_TA_UCCExample',
            can_change_perms: true,
            can_list: true,
            can_share_app: true,
            can_share_global: true,
            can_share_user: true,
            can_write: true,
            modifiable: true,
            owner: 'admin',
            perms: {
                read: ['*'],
                write: ['admin', 'sc_admin'],
            },
            removable: true,
            sharing: 'global',
        },
        content: {
            account_multiple_select: 'one',
            account_radio: '1',
            auth_type: 'basic',
            custom_endpoint: 'login.example.com',
            disabled: false,
            'eai:acl': null,
            'eai:appName': 'Splunk_TA_UCCExample',
            'eai:userName': 'nobody',
            password: '******',
            custom_text: '222222',
            token: '******',
            username: 'zzzzz',
        },
    },
];

export const MockRowData = {
    links: {
        create: `/servicesNS/nobody/-/splunk_ta_uccexample_account/_new`,
    },
    updated: '2023-08-21T11:54:12+00:00',
    entry: ROW_DATA,
    messages: [],
};

export const ServerHandlers = [
    http.get(`/servicesNS/nobody/-/splunk_ta_uccexample_account`, () =>
        HttpResponse.json(MockRowData)
    ),
];
