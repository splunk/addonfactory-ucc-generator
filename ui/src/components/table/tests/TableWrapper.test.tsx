import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import React from 'react';
import { http, HttpResponse } from 'msw';
import { BrowserRouter } from 'react-router-dom';
import { MockRowData } from '../stories/rowDataMockup';
import TableWrapper, { ITableWrapperProps } from '../TableWrapper';
import { server } from '../../../mocks/server';
import { TableContextProvider } from '../../../context/TableContext';
import { setUnifiedConfig } from '../../../util/util';
import { SIMPLE_NAME_TABLE_MOCK_DATA, TABLE_CONFIG_WITH_MAPPING } from '../stories/configMockups';

jest.mock('immutability-helper');

const handleRequestModalOpen = jest.fn();
const handleOpenPageStyleDialog = jest.fn();

it('correct render table with all elements', async () => {
    const props = {
        page: 'configuration',
        serviceName: 'account',
        handleRequestModalOpen,
        handleOpenPageStyleDialog,
        displayActionBtnAllRows: false,
    } satisfies ITableWrapperProps;

    server.use(
        http.get('/servicesNS/nobody/-/splunk_ta_uccexample_account', () =>
            HttpResponse.json(MockRowData)
        )
    );

    setUnifiedConfig(SIMPLE_NAME_TABLE_MOCK_DATA);

    render(
        <TableContextProvider>
            <TableWrapper {...props} />
        </TableContextProvider>,
        { wrapper: BrowserRouter }
    );

    const numberOfItems = await screen.findByText('9 Items');
    expect(numberOfItems).toBeInTheDocument();

    const headerNames = ['Name', 'Actions'];

    const tableHeader = screen.getAllByRole('columnheader');

    expect(tableHeader.length).toEqual(headerNames.length);

    headerNames.forEach((name) => {
        const thWithName = Array.from(tableHeader).find(
            (thElem: HTMLElement) => thElem.dataset.testLabel === name
        );
        expect(thWithName).toBeTruthy();
    });

    const currentTab = SIMPLE_NAME_TABLE_MOCK_DATA.pages.configuration.tabs.find(
        (tab) => tab.name === props.serviceName
    );

    currentTab?.entity.forEach((confEntity) =>
        expect(screen.getByText(confEntity.label)).toBeInTheDocument()
    );
});

it('sort items after filtering', async () => {
    const props = {
        page: 'configuration',
        serviceName: 'account',
        handleRequestModalOpen,
        handleOpenPageStyleDialog,
        displayActionBtnAllRows: false,
    } satisfies ITableWrapperProps;

    server.use(
        http.get('/servicesNS/nobody/-/splunk_ta_uccexample_account', () =>
            HttpResponse.json(MockRowData)
        )
    );

    setUnifiedConfig(TABLE_CONFIG_WITH_MAPPING);

    render(
        <TableContextProvider>
            <TableWrapper {...props} />
        </TableContextProvider>,
        { wrapper: BrowserRouter }
    );

    const numberOfItems = await screen.findByText('Custom Text');
    expect(numberOfItems).toBeInTheDocument();

    const customHeader = document.querySelector('[data-test-label="Custom Text"]');
    expect(customHeader).toBeInTheDocument();

    const defaultOrder = document.querySelectorAll('[data-column="custom_text"]');
    const mappedTextDefaultOrder = Array.from(defaultOrder).map((el: Node) => el.textContent);
    expect(mappedTextDefaultOrder).toMatchInlineSnapshot(`
        [
          "wxyz=a",
          "xyz=ab",
          "yz=abc",
          "z=abcd",
          "xyz=ab",
          "aaaaa",
          "two",
          "testsomethingelse",
          "222222",
        ]
    `);

    await userEvent.click(customHeader!);

    const allCustomTextsAsc = document.querySelectorAll('[data-column="custom_text"]');
    const mappedTextAsc = Array.from(allCustomTextsAsc).map((el: Node) => el.textContent);

    expect(mappedTextAsc).toMatchInlineSnapshot(`
        [
          "222222",
          "aaaaa",
          "testsomethingelse",
          "two",
          "wxyz=a",
          "xyz=ab",
          "xyz=ab",
          "yz=abc",
          "z=abcd",
        ]
    `);

    await userEvent.click(customHeader!);

    const allCustomTextsDesc = document.querySelectorAll('[data-column="custom_text"]');
    const mappedTextDesc = Array.from(allCustomTextsDesc).map((el: Node) => el.textContent);

    expect(mappedTextDesc).toMatchInlineSnapshot(`
        [
          "z=abcd",
          "yz=abc",
          "xyz=ab",
          "xyz=ab",
          "wxyz=a",
          "two",
          "testsomethingelse",
          "aaaaa",
          "222222",
        ]
    `);
});
