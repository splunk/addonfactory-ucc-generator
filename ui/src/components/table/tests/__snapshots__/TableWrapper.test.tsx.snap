// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sort items after filtering 1`] = `
[
  "wxyz=a",
  "xyz=ab",
  "yz=abc",
  "z=abcd",
  "xyz=ab",
  "aaaaa",
  "two",
  "testsomethingelse",
  "222222",
]
`;

exports[`sort items after filtering 2`] = `
[
  "222222",
  "aaaaa",
  "testsomethingelse",
  "two",
  "wxyz=a",
  "xyz=ab",
  "xyz=ab",
  "yz=abc",
  "z=abcd",
]
`;

exports[`sort items after filtering 3`] = `
[
  "z=abcd",
  "yz=abc",
  "xyz=ab",
  "xyz=ab",
  "wxyz=a",
  "two",
  "testsomethingelse",
  "aaaaa",
  "222222",
]
`;
